testing/matrix-sdk-test/src/event_builder.rs

```diff
@@ -0,0 +1,207 @@
+// Copyright 2023 The Matrix.org Foundation C.I.C.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Unit tests (based on private methods) for the timeline API.
+
+use std::{
+    collections::BTreeMap,
+    sync::atomic::{AtomicU64, Ordering::SeqCst},
+};
+
+use ruma::{
+    events::{
+        receipt::{Receipt, ReceiptEventContent, ReceiptThread, ReceiptType},
+        relation::Annotation,
+        AnySyncTimelineEvent, AnyTimelineEvent, MessageLikeEventContent,
+        RedactedMessageLikeEventContent, RedactedStateEventContent, StateEventContent,
+    },
+    serde::Raw,
+    server_name, EventId, MilliSecondsSinceUnixEpoch, OwnedEventId, OwnedUserId, RoomId, UserId,
+};
+use serde_json::{json, Value as JsonValue};
+
+#[derive(Default)]
+pub struct EventBuilder {
+    next_ts: AtomicU64,
+}
+
+impl EventBuilder {
+    pub fn new() -> EventBuilder {
+        Self { next_ts: AtomicU64::new(0) }
+    }
+
+    pub fn next_server_ts(&self) -> MilliSecondsSinceUnixEpoch {
+        MilliSecondsSinceUnixEpoch(
+            self.next_ts
+                .fetch_add(1, SeqCst)
+                .try_into()
+                .expect("server timestamp should fit in js_int::UInt"),
+        )
+    }
+
+    /// Set the next server timestamp.
+    ///
+    /// Timestamps will continue to increase by 1 (millisecond) from that value.
+    pub fn set_next_ts(&self, value: u64) {
+        self.next_ts.store(value, SeqCst);
+    }
+
+    pub fn make_message_event_with_id<C: MessageLikeEventContent>(
```

Comment:
Overall design question: would it make sense to have one or two methods here that return builders that do follow a Builder pattern, so we don't have many small methods focused on a single thing, and can have more composability? (I would imagine the `EventBuilder` name would need to be replaced, as the type returned by those methods would be an `EventBuilder`)

Reply:
Postponing this to a follow-up PR.